crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -169,22 +169,38 @@ impl<'db> IntersectionBuilder<'db> {
             // (T2 & T4)`. If `self` is already a union-of-intersections `(T1 & T2) | (T3 & T4)`
             // and we add `T5 | T6` to it, that flattens all the way out to `(T1 & T2 & T5) | (T1 &
             // T2 & T6) | (T3 & T4 & T5) ...` -- you get the idea.
-            union
+            return union
                 .elements(self.db)
                 .iter()
                 .map(|elem| self.clone().add_positive(*elem))
                 .fold(IntersectionBuilder::empty(self.db), |mut builder, sub| {
                     builder.intersections.extend(sub.intersections);
                     builder
-                })
-        } else {
-            // If we are already a union-of-intersections, distribute the new intersected element
-            // across all of those intersections.
-            for inner in &mut self.intersections {
-                inner.add_positive(self.db, ty);
+                });
+        }
+
+        if let Type::TypeVar(typevar) = ty {
+            if let Some(TypeVarBoundOrConstraints::Constraints(constraints)) =
+                typevar.bound_or_constraints(self.db)
+            {
+                // Ditto for the union of constraints in a constrained typevar.
+                return constraints
+                    .elements(self.db)
+                    .iter()
+                    .map(|constraint| self.clone().add_positive(*constraint))
+                    .fold(IntersectionBuilder::empty(self.db), |mut builder, sub| {
+                        builder.intersections.extend(sub.intersections);
+                        builder
+                    });
```

Comment:
As discussed in comments on the tests, I'm not convinced that this gloss of a constrained typevar to a union of its constraints is something that we should be doing here. Or at least -- I think we need to limit it to the case where we are immediately able to eliminate all but one of the union elements, so the result is no longer a union. If that's tricky, I also think it's OK to leave the entire ability to narrow a constrained typevar as a follow-up item instead.